### DISCUSSION TITLE
The difference between forking and cloning a repository on GitHub

### DISCUSSION THREAD

**Discussion by fernandez** [community/community#35849] [Discover: GitHub Best Practices]
# Cloning and forking repositories on GitHub

One of the main benefits of using a version control system is the ability to acquire your own copy of a project quickly and easily. The two most commonly used methods for copying a repository on GitHub are [cloning](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) and [forking](https://docs.github.com/en/get-started/quickstart/fork-a-repo).

## What is cloning?

Repositories created on GitHub exist as [remotes](https://docs.github.com/en/get-started/getting-started-with-git/about-remote-repositories). When you clone a repository you are creating a local copy on your computer that you can sync with the remote on GitHub. Cloning is ideal for contributing directly to a repository alongside other users while utilizing [branching](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-branches) and other collaboration tools to manage changes.

##### Tip: Cloning a repository can be used to create a [backup](https://docs.github.com/en/repositories/archiving-a-github-repository/backing-up-a-repository) that is regularly kept up to date, but it's important to note that although your clone will copy over Git data like files and commit history, it won't bring over issues, pull requests, and other GitHub elements.

To clone, head to the main page of a repository and click the green `Code` button. In the dropdown that appears you will have the option to clone over [HTTPS](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository#cloning-a-repository), [SSH](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/about-ssh), or by using the [GitHub CLI](https://cli.github.com/). 

## What about forking?

A fork is a copy of a repository that allows you to make your own changes without impacting the original project. A fork differs from a cloned copy in that it doesn't allow for direct collaboration with the root using local commands like `git push` and `git pull`. Instead, your fork exists on GitHub and you can contribute back to the original project using [Pull Requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests). You can also [synch your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) easily to keep it up-to-date with changes from the root repository.

Forks are ideal for situations where the root repository is serving as a basis for further iteration, or to play around with ideas -- instead of starting a project from scratch you can use an existing repository as a foundation then take it to the next level!

##### Tip: Keep in mind when working with a fork of a private repository that GitHub will consider the owner of the root repository to have ownership of its forks as well. This means that when a private repository is deleted any of its forks will also be removed. This isn't something to worry about for public repositories as in that case the fork is owned by the fork owner and will be detached or re-routed in the case a root is deleted.

Forking a project is as easy as clicking the `Fork` button in the header of the repository home page. Once the process is complete you'll be taken to your forked copy where you can start collaborating! Forking a repository will copy data such as files and code but Issues, branches, pull requests and other features won't join the party. Instead, your fork will start the same way as a newly created repository so you can begin work on it as a fresh project.

## Further Reading

Now that you know which method to use when making a new copy of a project, here are some other resources for getting comfortable working with forks and clones:

* [Understanding the GitHub Flow](https://docs.github.com/en/get-started/quickstart/github-flow)
* [Configuring a remote for a fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-for-a-fork)
* [Which remote URL should I use?](https://docs.github.com/en/get-started/getting-started-with-git/about-remote-repositories)

**Comment by mm7703** [comment]
**H hey guys**

@zackfern _[![image](https://user-images.githubusercontent.com/113254979/195934998-68028acd-7a75-4539-96ba-fba5b43be489.jpeg)]()_

**Reply by mm7703** [reply]
- - ![image](https://user-images.githubusercontent.com/113254979/195935144-77cd2654-54b9-4ffd-b119-921ad0107dd3.png)

**Comment by shumalov0** [comment]
Hi guys

**Comment by SBalint2002** [comment]
Hi guys

**Comment by S1983tt** [comment]
go to strip  GO GO

**Comment by Preetiraj3697** [comment]
Anyone give me suggestion how to I achieve more achievement in GitHub.

**Comment by mdriyadkhaan** [comment]
Good

**Reply by Zanati21** [reply]
gh pr checkout 34146

**Reply by kolumnin** [reply]
2022 October 30 06:38@kolumnin

**Reply by Alseed8889** [reply]
All files have been deleted from the device I want to restore what has been deleted

**Reply by miraclecustom** [reply]
just join it @miraclecustom

**Reply by Talalgmel** [reply]
Hi

**Reply by Talalgmel** [reply]
Hi

**Reply by 7und3dpy** [reply]
gud luck

**Reply by kolumnin** [reply]
Thank you.

**Comment by savaonepunch** [comment]
So if I want to contribute to a random repo, do I have to fork it and create a pull request? Or can I just clone it and create a new fork on the original repo?

**Reply by dserodio** [reply]
If you intend to contribute, you should fork it.

**Reply by souvikrajsingh** [reply]
> So if I want to contribute to a random repo, do I have to fork it and create a pull request? Or can I just clone it and create a new fork on the original repo?

if you want to contribute to a repo, you first fork it to your profile and the you clone it, after you push the changes to your branch you'll get the option to make a PR appearing on your forked repo! Simple

**Reply by Alseed8889** [reply]
تم حذف جميع الملفات من الجهاز اريد استعادة ماتم حذفة اهل استطيع

**Reply by Gainpay** [reply]
Pls I just join , I need someone to guide me

**Reply by cx0y** [reply]
if you wanna test some projects,  just clone its repository and if you wanna contribute to some repository, best practice is, create a fork, then commit your contribution and create a pull request to its upstream!

**Reply by om4rAb** [reply]
Let's say I am working on a restaurant project with some guys, So I need to fork it then clone it and do the work then push to the repo I forked then make a pull request to the originale repo right ?

**Reply by SaadHusayn** [reply]
yes you are right, but if you just want to test the project, you clone the original repo and work with that.

**Comment by varma59** [comment]


**Reply by frzyb** [reply]
no

**Reply by dserodio** [reply]
Not "genuine owner", but _original author_. The original author should be credited, you shouldn't "hide" the origin of the code. Even if the license allows, it's bad manners.

**Comment by imran-baitham** [comment]
> One of the main benefits of using a version control system is the ability to acquire your own copy of a project quickly and easily. The two most commonly used methods for copying a repository on GitHub are [cloning](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) and [forking](https://docs.github.com/en/get-started/quickstart/fork-a-repo).
>

**Comment by Bangden4890** [comment]
Saya tersesat

**Reply by Bangden4890** [reply]
Tolong ajari saya jadi heker

**Comment by Moam5700d** [comment]
I sm living in iran

**Comment by arijhajji-1** [comment]
thank you

**Reply by Alseed8889** [reply]
How to restore what has been deleted in the device through the application

**Comment by JPTbohs** [comment]
> **H hey guys**
> 
> @zackfern _![image](https://user-images.githubusercontent.com/113254979/195934998-68028acd-7a75-4539-96ba-fba5b43be489.jpeg)_

Stop all this link for now

**Comment by Alseed8889** [comment]
يوجد حل استعادة ماتم حذفة في الجهاز تم الإرسال من هاتف Huawei الخاص بي-------- الرسالة الأصلية --------من:‏ JPTbohs ***@***.***>التاريخ:‏ الاثنين، ٣١ أكتوبر، ٢٠٢٢ ٦:٤٣ مإلى:‏ community/community ***@***.***>نسخة إلى:‏ Alseed8889 ***@***.***>, Comment ***@***.***>الموضوع:‏ Re: [community/community] The difference between forking and cloning a repository on GitHub (Discussion #35849)

H hey guys
@zackfern 

Stop all this link for now

**Reply by Alseed8889** [reply]
سHow to restore what has been deleted in the device

**Comment by Alseed8889** [comment]
People I can get back. What has been deleted in the device تم الإرسال من هاتف Huawei الخاص بي-------- الرسالة الأصلية --------من:‏ arijhajji ***@***.***>التاريخ:‏ الاثنين، ٣١ أكتوبر، ٢٠٢٢ ٢:٤٤ مإلى:‏ community/community ***@***.***>نسخة إلى:‏ Alseed8889 ***@***.***>, Comment ***@***.***>الموضوع:‏ Re: [community/community] The difference between forking and cloning a repository on GitHub (Discussion #35849)
thank you

**Comment by Alseed8889** [comment]
تسطيع مساعدتي تطبيق SD Maid خذف جميع الملفات الصور والفيديو الصوت ابغى استعادة ماتم حذفة تم الإرسال من هاتف Huawei الخاص بي-------- الرسالة الأصلية --------من:‏ miraclecustom ***@***.***>التاريخ:‏ الاثنين، ٣١ أكتوبر، ٢٠٢٢ ١٠:٣٣ صإلى:‏ community/community ***@***.***>نسخة إلى:‏ Alseed8889 ***@***.***>, Comment ***@***.***>الموضوع:‏ Re: [community/community] The difference between forking and cloning a repository on GitHub (Discussion #35849)
just join it @miraclecustom

**Comment by Tinoet** [comment]
Good read. Thanks.

**Comment by 3kh0** [comment]
Thank you for this post, I wanted to ask if #38021 is on the roadmap and could you please add it if possible?

**Comment by SPerju** [comment]
Interesting

**Comment by Shaker-Al-Muzaini** [comment]
.

**Comment by Shaker-Al-Muzaini** [comment]
Thank you for this post

**Comment by MoussaAmbassador** [comment]
Thanks for sharing

**Comment by airyadriana** [comment]
The article says: Tip: Keep in mind when working with a fork of a private repository that GitHub will consider the owner of the root repository to have ownership of its forks as well. This means that when a private repository is deleted any of its forks will also be removed. This isn't something to worry about for public repositories as in that case the fork is owned by the fork owner and will be detached or re-routed in the case a root is deleted.

Question: We just created an organization in which a particular team wants to work with another organization's existing private repository (repo A) to improve their code so we don't start from scratch. However, we don't want to contribute back to repo A. My concern is that this article mentions that if repo A gets deleted, then our fork (if we fork) will be removed. Should we clone it to our Organization's team's repository instead?
Thanks!

**Comment by Shifat7** [comment]
Scenario: What if the OG owner's repo I forked from was public at first then they went private. What will happen to my fork. And if they delete the repo soon after making it private will my fork also get deleted? @fernandez

**Comment by fernandez** [comment]
Hello @Shifat7 👋 Thanks for your question. When a public repository is made private, any existing forks will be automatically 'detached' and setup in a new fork network that will remain public and available. If it is helpful, more details around what happens to forks when a repository's visibility changes can be found here:

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/what-happens-to-forks-when-a-repository-is-deleted-or-changes-visibility#changing-a-public-repository-to-a-private-repository

**Comment by fernandez** [comment]
Hi @airyadriana 👋 -- apologies for missing your question to this post. In case this information is still helpful to you and your team, it is correct that in the case you outlined (where you **do not** want to contribute back up to the main repository) the optimal strategy would be to clone the repository rather than forking it so that no changes to the root repo will impact your copy.

**Comment by kuruvinetwork-maker** [comment]
How to clone repo
